### DISCUSSION TITLE
2021-09-16 (automatic release)

### DISCUSSION THREAD

**Discussion by pipcet** [pipcet/pearl#25] [General]
(There was briefly a release, but then I realized my code to upload the release assets hadn't worked. So I'll just post this while waiting for a (fixed, hopefully) release run to complete...)

This release should fix the framebuffer error seen when terminating X or using non-X windowing systems. It does so rather crudely, but it's a temporary solution. It also adds support for zstd in the initial pearl environment, which is currently unused but will replace xz soon, and should make things faster when booting Debian.

Details and instructions in the discussion!

**Comment by pipcet** [comment]
For some reason the automated asset upload code isn't working, so I ended up manually uploading the build artifact as a release asset. Sorry about the delay.

**Comment by pipcet** [comment]
Instructions

This release fixes a number of bugs, including, hopefully, the crash we've usually seen when X terminates. The installer should continue to work. There's now a short https://bit.ly URL so there's less to type, too (thanks again!). Please forgive the optimism both here and in the rest of this post. Not everything might work, and please let me know if it doesn't.

To install pearl, boot into Recovery Mode, select Options, then Terminal from the Utilities menu, then enter

```
curl -L https://bit.ly/getpearl|perl
```

Please make sure that you're using HTTPS, as you're feeding a perl script directly into a perl interpreter running as the root user. (It would be a good idea to verify the integrity of the file with `sha512sum`, but that's not available in the Recovery Mode environment. `md5` is, but that's no longer a secure choice.)

Alternatively, download the pearl.pl script from https://github.com/pipcet/pearl/releases/latest/download/pearl.pl, verify its integrity, copy it to a location accessible from Recovery Mode, and run it through perl.

After the download completes, you will be prompted for nvram options for pearl. It is a good idea to specify "root /dev/mapper/debian-root" or something similar here, even if that partition does not exist, as it will save you a reboot into Recovery Mode lateron.

After that, make sure you understand what you are doing and, if you decide to go ahead, enter your username and password. This is as good a time as any to remind you there are no warranties and you're doing this at your own risk.

The screen comes back on with a full-screen text console and some text scrolls by.  Do nothing and wait.

The screen comes on again with four terminal windows: the upper left is the menu for interaction, the upper right is a general help/system status window, the lower left is for kernel messages, the lower right is for a shell prompt. There's an uptime counter and a countdown to when Debian will boot automatically. Do nothing and wait.

The screen will freeze for a few seconds and come back on with a full-screen text console. Many messages appear, some of which may sound worrying but are most likely harmless. If you are prompted for a passphrase (multiple times), hit Ctrl-D to skip the prompt (note that the prompt may be hidden between other messages); otherwise, do nothing and wait.

The Debian initrd is uncompressed only now, and it uses a comparatively slow compression algorithm, so there will be another delay. Do nothing and wait.

If you're on a mini, we now attempt to switch to 4k@60Hz resolution. That might look a bit strange as the framebuffer is switched. The screen goes dark for about ten seconds, then comes back on with a high-resolution text console. Do nothing and wait.

If you've selected a root file system, the installer should start automatically after another delay. Follow its instructions. Otherwise, the `root` user will be logged in automatically and `nmtui`, the NetworkManager interactive curses application, will be started. Select "Activate a connection"; both WiFi networks and the wired ethernet network should appear as appropriate for your device, so you can configure a network connection.

Once you quit `nmtui`, ensure you can reach the internet.

At this point, or later, you can enter "installer" at the shell prompt to launch Debian's graphical installer. In the installer, following the instructions should generally work. In order to boot automatically, create an LVM volume group called `debian` and a volume called `root` (or anything else you specified at the initial prompt; note that the physical volume may be encrypted, and you will be prompted for a decryption password automatically). Root partitions using the btrfs file system require specifying `root /dev/mapper/debian-root -o subvol=@rootfs` or similar at the initial prompt. It is expected that there is a warning about the lack of a known bootloader.

You're now logged in to the `root` account of a live Debian system running from a RAM filesystem: that means any changes you make locally will be lost as soon as you reboot. The RAM file system is initially limited in size. You can increase that by running `mount -o remount,size=7G /`, substituting any reasonable size for "7G".

You can now install, for example, a graphical desktop environment by running:
```
apt-get clean
apt-get update
apt-get install task-gnome-desktop
apt-get clean
init 5
```

To log in, use the non-root user that was created for you: username and password are both "y".

Finally, I'd be happy to hear from you! Please feel free to respond right in this discussion, open an issue, or contact me by email if you have any thoughts, observations, screenshots, problems, comments, or suggestions (or anything else, really).

Technical details:
- both mac addresses are chosen at random rather than set to the physical address.
- USB keyboards connected to the mini's Type A slot should work in the final Debian system, but will not work in the Pearl menu. USB keyboards connected to the Type C ports will.

**Comment by Johnnie390** [comment]
Pip,

just tested this morning. With the NVRAM options **no4k;root /dev/nvme0n1p4** does not seem to work i.e. black screen, no diagnostics etc. (I presume something died/went pear-shaped).

With the NVRAM options no4k on its own, I am able to reach a desktop after issuing init 5, (Gnome). After approximately 5 seconds everything shuts down, please see photo.
![Photo-I](https://user-images.githubusercontent.com/22815859/133760604-593edb31-b472-4ab7-8c6e-4ef10cb79648.jpeg)

To rule out the Gnome desktop as culprit, I started over again and installed (task-lxqt-desktop) an LXQT desktop with more or less identical results as the Gnome desktop.. Photo 
![Photo-II](https://user-images.githubusercontent.com/22815859/133760889-c7ccde92-fe6b-4338-a0e9-06f790380620.jpeg)

Regards,

Ry

**Reply by pipcet** [reply]
Thanks for testing! As for the second problem, it looks like it's shutting down in the regular fashion. I have two theories
1. the power button code might have generated a spurious power button event
2. the battery code might have returned a reading of 0, and systemd might have shut down the system in response to this empty battery phantom reading

As for the first problem, is the black screen happening immediately after reboot? That is very strange, because it's before the nvram variable is even read.

I'll be able to do some more testing later to see whether I can reproduce these issues. If you want to try something, can you try launching the installer instead of GNOME, to see whether it shuts down even then? (IIUC, the installer doesn't use systemd, so that would be a hint one of my theories might be correct).

**Comment by pipcet** [comment]
2021-09-17 is out, but it probably doesn't fix the issues @Johnnie390 reported. More testing later.
